lib/helpers.js

```diff
@@ -0,0 +1,144 @@
+const _ = require('lodash');
+
+function resolveAuthLevel(schema, options, doc) {
+  // Look into options the options and try to find authLevels. Always prefer to take
+  // authLevels from the direct authLevel option as opposed to the computed
+  // ones from getAuthLevel in the schema object.
+  let authLevels = [];
+  if (options) {
+    if (options.authLevel) {
+      authLevels = _.castArray(options.authLevel);
+    } else if (typeof schema.getAuthLevel === 'function') {
+      authLevels = _.castArray(schema.getAuthLevel(options.authPayload, doc));
+    }
+  }
+  // Add `defaults` to the list of levels since you should always be able to do what's specified
+  // in defaults.
+  authLevels.push('defaults');
+
+  const perms = schema.permissions || {};
+  return _.chain(authLevels)
+    .filter(level => !!perms[level]) // make sure the level in the permissions dict
+    .uniq() // get rid of fields mentioned in multiple levels
+    .value();
+}
+
+function getAuthorizedFields(schema, options, action, doc) {
+  const authLevels = resolveAuthLevel(schema, options, doc);
+
+  return _.chain(authLevels)
+    .flatMap(level => schema.permissions[level][action])
+    .filter(path => schema.path(path) || schema.virtualpath(path)) // ensure fields are in schema
```

Comment:
This line breaks access to subdocuments for me (posted a failing test on the forked repo), is there a security reason that we can't access things that aren't in the schema?

Reply:
Let's chat about this in a new issue. I just created #8 for us to discuss.